### ISSUE TITLE
Suggest rewrite `arr.iter.all(f)` to easy-to-unroll form under very-limited condition

### ISSUE THREAD

**Issue by KisaragiEffective** [rust-lang/rust-clippy#11354]
### What it does

Suggests rewriting logic if all following condition holds;
* `Iterator` is created from array or reference to it
* creation of iterator is followed by `Iterator::all`
* the lambda of `Iterator::all` is clearly Disjunctive normal form where each conjunction only compare to literal

### Advantage

- Improves execution time
  - Interestingly, the naive `arr.iter().all(|x| f(x))` is not unrolled as of 1.71.0

### Drawbacks

- Increases code size
  - In release mode, it will attempt to optimized
- May difficult to read
- Also may be too niche to have
  - So this lint would be fall to restriction category
- Concern: do this holds under 32-bit or 16-bit targets?

### Example

```rust
const N: usize = 256;

pub fn co1(x: &u8) -> bool {
    *x == 0x21
}

pub fn co2(x: &u8) -> bool {
    *x <= 0x08
}

pub fn co3(x: &u8) -> bool {
    *x >= 0x26
}

// here
pub fn check_or_chain(x: [u8; N]) -> bool {
    x.iter().all(|b| co1(b) || co2(b) || co3(b))
}
```

Could be written as:

```rust
pub fn check_or_chain(x: [u8; N]) -> bool {
    x.iter().filter(|b| co1(*b)).count() +
    x.iter().filter(|b| co2(*b)).count() + 
    x.iter().filter(|b| co3(*b)).count() + 
    0
    == N
}
```

**Comment by Alexendoo** [comment]
I don't think we should lint specific compilation inefficiencies, those are a very moving target and we have no real way to test that they stay inefficient

They're better off as an issue in the main rust repo

**Comment by Centri3** [comment]
Your "good" code is also 2x slower; 160ns -> 372ns on nightly. Did you get the order mixed up?

See the [assembly](https://rust.godbolt.org/z/8sY4d7zr8), it's the same on 1.71.0 as well
